awx/main/tasks/jobs.py

```diff
@@ -570,18 +567,20 @@ def run(self, pk, **kwargs):
         except PostRunError as exc:
             if status == 'successful':
                 status = exc.status
-                extra_update_fields['job_explanation'] = exc.args[0]
+                self.runner_callback.delay_update(job_explanation=exc.args[0])
                 if exc.tb:
-                    extra_update_fields['result_traceback'] = exc.tb
+                    self.runner_callback.delay_update(result_traceback=exc.tb)
         except Exception:
             logger.exception('{} Post run hook errored.'.format(self.instance.log_format))
 
         # We really shouldn't get into this one but just in case....
-        if 'got an unexpected keyword argument' in extra_update_fields.get('result_traceback', ''):
-            extra_update_fields['result_traceback'] = "{}\n\n{}".format(extra_update_fields['result_traceback'], ANSIBLE_RUNNER_NEEDS_UPDATE_MESSAGE)
+        original_traceback = self.runner_callback.get_extra_update_fields().get('result_traceback', '')
+        if 'got an unexpected keyword argument' in original_traceback:
+            if ANSIBLE_RUNNER_NEEDS_UPDATE_MESSAGE not in original_traceback:
+                self.runner_callback.delay_update(result_traceback=ANSIBLE_RUNNER_NEEDS_UPDATE_MESSAGE)
```

Comment:
One of the things I never really loved was this code being somewhat duplicated in here and receptor.py could we just move this into an if statement in delay_update?

Reply:
Probably. What I think that would look like - if `delay_update` gets `job_explanation` or `result_traceback` and the given string is already in that field, then it does nothing. So we would never repeat the same string twice. Seems reasonable but I should add a test.